File: guides/export-pdf.py
#!/usr/bin/env python3
"""Export Markdown guides to PDF with inline LaTeX math ($...$).

Pipeline: Markdown → python-markdown + markdown-katex (KaTeX HTML/SVG)
→ WeasyPrint PDF.

Requires the project venv (see guides/requirements-pdf.txt):
  python3 -m venv .venv-pdf && .venv-pdf/bin/pip install -r guides/requirements-pdf.txt
  .venv-pdf/bin/python guides/export-pdf.py [input.md] [output.pdf]
"""

from __future__ import annotations

import argparse
import re
import sys
from pathlib import Path

import markdown
from markdown.extensions.fenced_code import FencedCodeExtension
from markdown.extensions.tables import TableExtension
from markdown_katex.extension import KatexExtension
from weasyprint import CSS, HTML

GUIDES_DIR = Path(__file__).resolve().parent
DEFAULT_MD = GUIDES_DIR / "AI课程-14周内容梳理.md"
DEFAULT_PDF = GUIDES_DIR / "AI课程-14周内容梳理.pdf"
CSS_FILE = GUIDES_DIR / "pdf-export.css"


def prepare_blockquotes(text: str) -> str:
    """Insert blank blockquote lines before lists inside ``>`` blocks.

    python-markdown treats ``> intro:\\n> - item`` as one paragraph unless a
    blank ``>`` line separates the intro from the list.
    """
    lines = text.splitlines()
    result: list[str] = []
    bq = re.compile(r"^>\s?(.*)$")
    list_start = re.compile(r"^(\s*)([-*+]|\d+\.)\s")

    i = 0
    while i < len(lines):
        line = lines[i]
        match = bq.match(line)
        if match and i + 1 < len(lines):
            nxt = bq.match(lines[i + 1])
            if nxt:
                cur_content = match.group(1)
                nxt_content = nxt.group(1)
                cur_is_list = bool(list_start.match(cur_content))
                nxt_is_list = bool(list_start.match(nxt_content))
                if not cur_is_list and nxt_is_list and cur_content.strip():
                    result.append(line)
                    result.append(">")
                    i += 1
                    continue
        result.append(line)
        i += 1
    return "\n".join(result)


# Color emoji embed as raster images in WeasyPrint and leave green corner artifacts.
PDF_SYMBOL_REPLACEMENTS = {
    "✅": "✓",
    "🏖️": "〔休〕",
    "🏃": "〔运〕",
    "🔜": "→",
}


def prepare_pdf_symbols(text: str) -> str:
    """Replace color emoji with plain-text symbols safe for WeasyPrint."""
    for src, dst in PDF_SYMBOL_REPLACEMENTS.items():
        text = text.replace(src, dst)
    return text


def prepare_markdown_lists(text: str) -> str:
    """Normalize list markup for python-markdown.

    - Insert a blank line before a top-level list when it immediately follows prose
      (e.g. ``**核心内容**：`` + ``- item``), otherwise the list is parsed as a paragraph.
    - Expand 2-space nested list indents to 4-space (``  -`` → ``    -``).
    """
    lines = text.splitlines()
    prepared: list[str] = []
    list_line = re.compile(r"^(\s*)([-*+]|\d+\.)\s")

    for line in lines:
        match = list_line.match(line)
        if match and match.group(1) == "" and prepared and prepared[-1].strip():
            if not list_line.match(prepared[-1]):
                prepared.append("")
        prepared.append(line)

    normalized: list[str] = []
    for line in prepared:
        match = list_line.match(line)
        if match and match.group(1):
            level = len(match.group(1)) // 2
            normalized.append(" " * (4 * level) + line[len(match.group(1)) :])
        else:
            normalized.append(line)
    return "\n".join(normalized)


def convert_dollar_math(text: str) -> str:
    """Map $...$ / $$...$$ to markdown-katex fenced syntax ($`...`$)."""
    text = re.sub(
        r"\$\$([^\$]+)\$\$",
        lambda m: "\n```math\n" + m.group(1).strip() + "\n```\n",
        text,
    )
    text = re.sub(
        r"\$([^\$\n]+)\$",
        lambda m: "$`" + m.group(1) + "`$",
        text,
    )
    return text


def prepare_markdown(md_text: str) -> str:
    return prepare_pdf_symbols(
        prepare_markdown_lists(prepare_blockquotes(md_text))
    )


def md_to_html(md_text: str) -> str:
    body = markdown.markdown(
        convert_dollar_math(prepare_markdown(md_text)),
        extensions=[
            KatexExtension(no_inline_svg=True),
            TableExtension(),
            FencedCodeExtension(),
        ],
    )
    return f"""<!DOCTYPE html>
<html lang="zh-CN">
<head>
  <meta charset="utf-8">
  <title>Guide Export</title>
</head>
<body>
{body}
</body>
</html>"""


def export_pdf(md_path: Path, pdf_path: Path) -> None:
    md_text = md_path.read_text(encoding="utf-8")
    html_doc = md_to_html(md_text)
    pdf_path.parent.mkdir(parents=True, exist_ok=True)
    HTML(string=html_doc, base_url=str(GUIDES_DIR)).write_pdf(
        str(pdf_path),
        stylesheets=[CSS(filename=str(CSS_FILE))],
    )


def main(argv: list[str] | None = None) -> int:
    parser = argparse.ArgumentParser(description="Export Markdown guide to PDF.")
    parser.add_argument(
        "input",
        nargs="?",
        type=Path,
        default=DEFAULT_MD,
        help=f"Source Markdown (default: {DEFAULT_MD.name})",
    )
    parser.add_argument(
        "output",
        nargs="?",
        type=Path,
        default=DEFAULT_PDF,
        help=f"Output PDF (default: {DEFAULT_PDF.name})",
    )
    args = parser.parse_args(argv)

    md_path = args.input.resolve()
    pdf_path = args.output.resolve()

    if not md_path.is_file():
        print(f"error: input not found: {md_path}", file=sys.stderr)
        return 1
    if not CSS_FILE.is_file():
        print(f"error: stylesheet not found: {CSS_FILE}", file=sys.stderr)
        return 1

    print(f"Exporting {md_path.name} → {pdf_path.name} …")
    export_pdf(md_path, pdf_path)
    print(f"Wrote {pdf_path} ({pdf_path.stat().st_size:,} bytes)")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: guides/export-pdf.mjs
#!/usr/bin/env node
/**
 * Export Markdown guides to PDF via Puppeteer (Chromium print).
 *
 * Pipeline: Markdown → marked (GFM) → HTML + KaTeX auto-render → page.pdf()
 *
 * Setup:
 *   cd guides && npm install
 *   node export-pdf.mjs [input.md] [output.pdf]
 */

import fs from "node:fs/promises";
import path from "node:path";
import process from "node:process";
import { fileURLToPath } from "node:url";
import { marked } from "marked";
import puppeteer from "puppeteer";

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const GUIDES_DIR = __dirname;
const DEFAULT_MD = path.join(GUIDES_DIR, "AI课程-14周内容梳理.md");
const DEFAULT_PDF = path.join(GUIDES_DIR, "AI课程-14周内容梳理.pdf");
const CSS_FILE = path.join(GUIDES_DIR, "pdf-export.css");

const KATEX_VERSION = "0.16.11";
const KATEX_CSS = `https://cdn.jsdelivr.net/npm/katex@${KATEX_VERSION}/dist/katex.min.css`;
const KATEX_JS = `https://cdn.jsdelivr.net/npm/katex@${KATEX_VERSION}/dist/katex.min.js`;
const KATEX_AUTO_RENDER = `https://cdn.jsdelivr.net/npm/katex@${KATEX_VERSION}/dist/contrib/auto-render.min.js`;

marked.setOptions({ gfm: true, breaks: false });

function prepareBlockquotes(text) {
  const lines = text.split("\n");
  const result = [];
  const bq = /^>\s?(.*)$/;
  const listStart = /^(\s*)([-*+]|\d+\.)\s/;

  for (let i = 0; i < lines.length; i += 1) {
    const line = lines[i];
    const match = bq.exec(line);
    if (match && i + 1 < lines.length) {
      const nxt = bq.exec(lines[i + 1]);
      if (nxt) {
        const curContent = match[1];
        const nxtContent = nxt[1];
        const curIsList = listStart.test(curContent);
        const nxtIsList = listStart.test(nxtContent);
        if (!curIsList && nxtIsList && curContent.trim()) {
          result.push(line, ">");
          continue;
        }
      }
    }
    result.push(line);
  }
  return result.join("\n");
}

function prepareMarkdownLists(text) {
  const lines = text.split("\n");
  const prepared = [];
  const listLine = /^(\s*)([-*+]|\d+\.)\s/;

  for (const line of lines) {
    const match = listLine.exec(line);
    if (match && match[1] === "" && prepared.length && prepared.at(-1).trim()) {
      if (!listLine.test(prepared.at(-1))) {
        prepared.push("");
      }
    }
    prepared.push(line);
  }

  const normalized = [];
  for (const line of prepared) {
    const match = listLine.exec(line);
    if (match && match[1]) {
      const level = Math.floor(match[1].length / 2);
      normalized.push(" ".repeat(4 * level) + line.slice(match[1].length));
    } else {
      normalized.push(line);
    }
  }
  return normalized.join("\n");
}

function prepareMarkdown(mdText) {
  return prepareMarkdownLists(prepareBlockquotes(mdText));
}

async function mdToHtmlDocument(mdText, cssText) {
  const body = marked.parse(prepareMarkdown(mdText));
  return `<!DOCTYPE html>
<html lang="zh-CN">
<head>
  <meta charset="utf-8">
  <title>Guide Export</title>
  <link rel="stylesheet" href="${KATEX_CSS}">
  <style>${cssText}</style>
</head>
<body>
${body}
</body>
</html>`;
}

async function exportPdf(mdPath, pdfPath) {
  const [mdText, cssText] = await Promise.all([
    fs.readFile(mdPath, "utf8"),
    fs.readFile(CSS_FILE, "utf8"),
  ]);
  const htmlDoc = await mdToHtmlDocument(mdText, cssText);

  await fs.mkdir(path.dirname(pdfPath), { recursive: true });

  const browser = await puppeteer.launch({
    headless: true,
    args: ["--no-sandbox", "--disable-setuid-sandbox"],
  });

  try {
    const page = await browser.newPage();
    await page.setContent(htmlDoc, { waitUntil: "networkidle0" });
    await page.addScriptTag({ url: KATEX_JS });
    await page.addScriptTag({ url: KATEX_AUTO_RENDER });
    await page.evaluate(() => {
      renderMathInElement(document.body, {
        delimiters: [
          { left: "$$", right: "$$", display: true },
          { left: "$", right: "$", display: false },
        ],
        throwOnError: false,
      });
    });
    await page.emulateMediaType("print");
    await page.pdf({
      path: pdfPath,
      format: "A4",
      printBackground: true,
      margin: { top: "20mm", right: "18mm", bottom: "20mm", left: "18mm" },
    });
  } finally {
    await browser.close();
  }
}

function parseArgs(argv) {
  const args = argv.slice(2);
  if (args.includes("-h") || args.includes("--help")) {
    return { help: true };
  }
  return {
    input: args[0] ? path.resolve(args[0]) : DEFAULT_MD,
    output: args[1] ? path.resolve(args[1]) : DEFAULT_PDF,
  };
}

async function main() {
  const { help, input: mdPath, output: pdfPath } = parseArgs(process.argv);
  if (help) {
    console.log(`Usage: node export-pdf.mjs [input.md] [output.pdf]

Default input : ${DEFAULT_MD}
Default output: ${DEFAULT_PDF}`);
    return 0;
  }

  try {
    await fs.access(mdPath);
  } catch {
    console.error(`error: input not found: ${mdPath}`);
    return 1;
  }
  try {
    await fs.access(CSS_FILE);
  } catch {
    console.error(`error: stylesheet not found: ${CSS_FILE}`);
    return 1;
  }

  console.log(`Exporting ${path.basename(mdPath)} → ${path.basename(pdfPath)} …`);
  await exportPdf(mdPath, pdfPath);
  const stat = await fs.stat(pdfPath);
  console.log(`Wrote ${pdfPath} (${stat.size.toLocaleString()} bytes)`);
  return 0;
}

main().then((code) => process.exit(code));
